AX: [AX Thread Text APIs] Downgrade release asserts in AXTextMarker

hoffmanjoshua opened the pull request:
#### 2f017d5b2c03fba704bb135506d9c2fbe78b9c9f
<pre>
AX: [AX Thread Text APIs] Downgrade release asserts in AXTextMarker
<a href="https://bugs.webkit.org/show_bug.cgi?id=289118">https://bugs.webkit.org/show_bug.cgi?id=289118</a>
<a href="https://rdar.apple.com/146146886">rdar://146146886</a>

Reviewed by Tyler Wilcock.

This PR downgrades the release asserts in AXTextMarker to debug, adds appropriate
fallbacks when these conditions fail, and adds new logging to help diagnose issues
when we hit them.

`accessibility/ax-thread-text-apis` and `accessibility/isolated-tree` do not change
their pass rate with this change.

* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::characterRange const):
(WebCore::AXTextMarker::lineIndex const):
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::markerRangeForLineIndex const):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::offsetFromRoot const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findLine const):
(WebCore::AXTextMarker::findParagraph const):
(WebCore::AXTextMarker::toTextRunMarker const):
(WebCore::AXTextMarker::partialOrderByTraversal const):
* Source/WebCore/accessibility/AXTextMarker.h:

Canonical link: <a href="https://commits.webkit.org/291629@main">https://commits.webkit.org/291629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14b2dbc2fc4782e8083a845d3dfb5eaa672124b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93447 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98448 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43973 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13301 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21463 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71386 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28770 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9947 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9630 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2135 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43287 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79915 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100479 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92113 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14993 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80402 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20751 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80433 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79731 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19832 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24272 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1620 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13625 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20483 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25661 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20170 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23630 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21911 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -569,7 +571,9 @@ CharacterRange AXTextMarker::characterRangeForLine(unsigned lineIndex) const
-    TEXT_MARKER_ASSERT(!offset());
+    TEXT_MARKER_ASSERT(!offset(), "characterRangeForLine");
+    if (offset())
+        return { };
```
I think it may be a better / closer-to-correct fallback to just continue on with the method as normal, rather than returning `{ }` early.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -547,7 +547,9 @@ int AXTextMarker::lineIndex() const
-        TEXT_MARKER_ASSERT_DOBULE(nextLineEndMarker.lineID() == currentMarker.lineID(), nextLineEndMarker, currentMarker);
+        TEXT_MARKER_ASSERT_DOUBLE(nextLineEndMarker.lineID() == currentMarker.lineID(), "lineIndex", nextLineEndMarker, currentMarker);
+        if (nextLineEndMarker.lineID() != currentMarker.lineID())
+            return -1;
```
If this condition fails, the worst that would happen is that the computed `index` is off-by-one, right? I feel like that's better than returning `-1` 🤔

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -598,7 +602,9 @@ CharacterRange AXTextMarker::characterRangeForLine(unsigned lineIndex) const
-    TEXT_MARKER_ASSERT(!offset());
+    TEXT_MARKER_ASSERT(!offset(), "markerRangeForLineIndex");
+    if (offset())
+        return { };
```
I think it may be a better / closer-to-correct fallback to just continue on with the method as normal, rather than returning { } early.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1011,7 +1023,10 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, CoalesceObjectBreak
     // offset() pointed to the last character in the given object's runs, so let's traverse to find the next object with runs.
     object = findObjectWithRuns(*object, direction, stopAtID);
     if (object) {
-        RELEASE_ASSERT(direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength());
+        bool nextRunHasLength = direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength();
+        TEXT_MARKER_ASSERT(nextRunHasLength, "findMarker");
+        if (!nextRunHasLength)
+            return { };
```
This is another case where I feel like just continuing on is maybe better than returning a completely invalid marker — plus this is duplicating some work (`direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength()` twice), which has a cost. What do you think?

hoffmanjoshua replied:
To your first point, I think this could lead to some invalid markers, because farther down we return:
```c++
        return AXTextMarker { *object, direction == AXDirection::Next ? startingOffset : object->textRuns()->lastRunLength() - startingOffset };
```
and if the `lastRunLength` is 0, we could end up overflowing if the offset is non-zero.

hoffmanjoshua replied:
In regards to duplicating work, my impression is that `nextRunHasLength` will calculate and store the resulting boolean once, and then we can use it twice without recalculating—but let me know if that is the wrong idea here.

twilco replied:
> In regards to duplicating work, my impression is that nextRunHasLength will calculate and store the resulting boolean once, and then we can use it twice without recalculating—but let me know if that is the wrong idea here.

Ah whoops, I misread, I thought we were doing the same thing as part of the `return` but that actually does something different, so ignore this one 😅

twilco commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -27,17 +27,31 @@
+    if (!(assertion)) \
+        TEXT_MARKER_LOG(methodName); \
     std::string debugString = "Text marker origin: " + originToString(origin()).utf8().toStdString(); \
-    RELEASE_ASSERT_WITH_MESSAGE(assertion, "%s", debugString.c_str()); \
+    ASSERT_WITH_MESSAGE(assertion, "%s", debugString.c_str()); \
```
I wonder if we should add `#if ASSERT_ENABLED` around all of the `std::string debugString` creations so we can be 100% we aren't paying the cost in release? Not sure how smart the compiler is.

```cpp
#if ASSERT_ENABLED \
std::string debugString = "Text marker origin: " + originToString(origin()).utf8().toStdString(); \
ASSERT_WITH_MESSAGE(assertion, "%s", debugString.c_str()); \
#endif \
```

hoffmanjoshua replied:
Yeah we can definitely do that!

hoffmanjoshua replied:
Scratch that—we can't use preprocessor flags inside macros

twilco reviewed (approved):
r+ with all comments addressed as discussed here and elsewhere

hoffmanjoshua committed:
AX: [AX Thread Text APIs] Downgrade release asserts in AXTextMarker
https://bugs.webkit.org/show_bug.cgi?id=289118
rdar://146146886

Reviewed by Tyler Wilcock.

This PR downgrades the release asserts in AXTextMarker to debug, adds appropriate
fallbacks when these conditions fail, and adds new logging to help diagnose issues
when we hit them.

`accessibility/ax-thread-text-apis` and `accessibility/isolated-tree` do not change
their pass rate with this change.

* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::characterRange const):
(WebCore::AXTextMarker::lineIndex const):
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::markerRangeForLineIndex const):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::offsetFromRoot const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::findLine const):
(WebCore::AXTextMarker::findParagraph const):
(WebCore::AXTextMarker::toTextRunMarker const):
(WebCore::AXTextMarker::partialOrderByTraversal const):
* Source/WebCore/accessibility/AXTextMarker.h:

Canonical link: https://commits.webkit.org/291629@main
